File: 566/reshape-the-matrix.go
func matrixReshape(nums [][]int, r int, c int) [][]int {
	var chanInt chan int
	chanInt = make(chan int)

	var length int
	go func(len *int) {
		for _, v1 := range nums {
			for _, v := range v1 {
				(*len)++
				chanInt <- v
			}
		}
		for {
			chanInt <- 0
		}
	}(&length)

	var res [][]int
	for i := 0; i < r; i++ {
		var row []int
		for j := 0; j < c; j++ {
			v := <-chanInt
			row = append(row, v)
		}
		res = append(res, row)
	}

	if r*c != length {
		return nums
	}

	return res
}

File: 566/reshape-the-matrix.py
class Solution:
    def matrixReshape(self, nums, r, c):
        """
        :type nums: List[List[int]]
        :type r: int
        :type c: int
        :rtype: List[List[int]]
        """
        rows, cols = len(nums), len(nums[0])
        if r * c != rows * cols:
            return nums

        res = [[0 for x in range(c)] for y in range(r)]
        for i in range(r):
            for j in range(c):
                k = i * c + j
                col_idx = k % cols
                row_idx = int((k - col_idx) / cols)
                res[i][j] = nums[row_idx][col_idx]

        return res
